File: csharp/day_01/Program.cs
using System.Collections.ObjectModel;

using Day_01.Services;


static (Collection<int> left, Collection<int> right) ReadFile(string path)
{
    Collection<int> leftRaw = [];
    Collection<int> rightRaw = [];

    IOService.ReadFile(path, (reader) =>
    {
        var line = reader.ReadLine();

        if (line != null)
        {
            var split = line.Split("   ");
            leftRaw.Add(int.Parse(split[0]));
            rightRaw.Add(int.Parse(split[1]));
        }
    });

    return (leftRaw, rightRaw);
}

static int CalculateTotalDistanceScore(Collection<int> left, Collection<int> right)
{
    var leftList = left.OrderBy(x => x).ToList();
    var rightList = right.OrderBy(x => x).ToList();
    var total = 0;

    for (int i = 0; i < leftList.Count; i++)
    {
        total += Math.Abs(leftList[i] - rightList[i]);
    }

    return total;
}

static int CalculateTotalSimalarityScore(Collection<int> left, Collection<int> right)
{
    var leftList = left.OrderBy(x => x).ToList();
    var rightList = right.GroupBy(x => x).ToDictionary(x => x.Key, x => x.Count());
    var total = 0;

    for (int i = 0; i < leftList.Count; i++)
    {
        var exists = rightList.TryGetValue(leftList[i], out var count);

        if (exists)
            total += leftList[i] * count;
    }

    return total;
}


var (left, right) = ReadFile("./data/location_id_list.txt");

var totalDistanceScore = CalculateTotalDistanceScore(left, right);
var totalSimalarityScore = CalculateTotalSimalarityScore(left, right);


Console.WriteLine("Total Distance score: {0}", totalDistanceScore);
Console.WriteLine("Total Simalarity Score: {0}", totalSimalarityScore);

File: fsharp/day_01/Program.fs
open Swensen.Unquote
open System
open System.IO

let getData path = 
    File.ReadAllLines(path)
    |> List.ofSeq

let parse (lines: string list) = 
    lines
    |> List.map (fun line -> line.Split("   ") |> Seq.map int |> Seq.toList)
    |> List.transpose

let order collection = 
    collection
    |> List.sort

let listToMap (collection: int list) =
    let addElement (acc:Map<int,int>) element = 
        match acc.TryGetValue(element) with
        | true, value -> acc.Add(element, value + 1)
        | _ -> acc.Add(element, 1)
    
    (Map.empty, collection)
    ||> Seq.fold (addElement)

let calculateTotalDistance (left : int list) (right : int list) = 
    List.zip left right
    |> List.sumBy (fun (l, r) -> abs (l - r))

let calculateTotalSimalarityScore (left : int list) (right : Map<int,int>) =
    left
    |> List.sumBy (fun l -> if right.ContainsKey l then l * right.[l] else 0)
    

let [left: int list; right: int list] = 
    getData "data/location_id_list.txt"
    |> parse

let leftList = order left
let rightList = order right
let rightMap = listToMap right


let totalDistanceScore = calculateTotalDistance leftList rightList
let totalSimalarityScore = calculateTotalSimalarityScore leftList rightMap

Console.WriteLine("Total Distance score: {0}", totalDistanceScore)
Console.WriteLine("Total Simalarity score: {0}", totalSimalarityScore)